third_party/nvidia/hopper/lib/Transforms/WarpSpecialization/TMEMAlloc1D.cpp

```diff
@@ -0,0 +1,229 @@
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Pass/PassManager.h"
+#include "nvidia/hopper/include/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonNvidiaGPU/IR/Dialect.h"
+
+#define DEBUG_TYPE "nvgpu-1D-tmem-alloc"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace tt = mlir::triton;
+namespace ttg = mlir::triton::gpu;
+namespace ttng = ::mlir::triton::nvidia_gpu;
+namespace mlir {
+
+// Wrapper class to hold the context for handling
+// 1D TMEM Allocation.
+class TMEM1DAllocator {
```

Comment:
I wrapped this in a class because otherwise other load needs to return multiple values (e.g. `tt::ExpandDimsOp` and `ttng::TMEMAllocOp`) in order to pass the information needed for layout conversions. Happy to refactor this, but this seemed nice and contained.